src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -0,0 +1,865 @@
+
+        double maxSize = controller.getMaxLevelDensity(0, controller.getBaseSstableSize(controller.getFanout(0)) / shardManager.localSpaceCoverage());
+        int index = 0;
+        Level level = new Level(controller, index, 0, maxSize);
+        for (SSTableReader candidate : suitable)
+        {
+            final double size = shardManager.density(candidate);
+            if (size < level.max)
+            {
+                level.add(candidate);
+                continue;
+            }
+
+            level.complete();
+            levels.add(level); // add even if empty
+
+            while (true)
+            {
+                ++index;
+                double minSize = maxSize;
+                maxSize = controller.getMaxLevelDensity(index, minSize);
+                level = new Level(controller, index, minSize, maxSize);
+                if (size < level.max)
+                {
+                    level.add(candidate);
+                    break;
+                }
+                else
+                {
+                    levels.add(level); // add the empty level
+                }
+            }
+        }
+
+        if (!level.sstables.isEmpty())
+        {
+            level.complete();
+            levels.add(level);
+        }
+
+        return levels;
+    }
+
+    private List<SSTableReader> getCompactableSSTables(Collection<SSTableReader> sstables,
+                                                       Predicate<SSTableReader> compactionFilter)
+    {
+        Set<SSTableReader> compacting = cfs.getTracker().getCompacting();
+        List<SSTableReader> suitable = new ArrayList<>(sstables.size());
+        for (SSTableReader rdr : sstables)
+        {
+            if (compactionFilter.test(rdr) && !compacting.contains(rdr))
+                suitable.add(rdr);
+        }
+        return suitable;
+    }
+
+    public TableMetadata getMetadata()
+    {
+        return cfs.metadata();
+    }
+
+    private static boolean startsAfter(SSTableReader a, SSTableReader b)
+    {
+        // Strict comparison because the span is end-inclusive.
+        return a.getFirst().compareTo(b.getLast()) > 0;
+    }
+
+    @Override
+    public String toString()
+    {
+        return String.format("Unified strategy %s", getMetadata());
+    }
+
+    /**
+     * A level: index, sstables and some properties.
+     */
+    public static class Level
+    {
+        final List<SSTableReader> sstables;
+        final int index;
+        final double survivalFactor;
+        final int scalingParameter; // scaling parameter used to calculate fanout and threshold
+        final int fanout; // fanout factor between levels
+        final int threshold; // number of SSTables that trigger a compaction
+        final double min; // min density of sstables for this level
+        final double max; // max density of sstables for this level
+        int maxOverlap = -1; // maximum number of overlapping sstables
```

Comment:
nit: Can you please add a line on why one would want to limit the overlap? What happens if a level/bucket crosses this threshold? Do we not run a compaction in that level? How do we recover?

Reply:
Added a line here and an explanation of the process in `getCompactionPick` as well as `getBuckets`.